File: OOP-Java/basics/com/private_constructor_demo/Car.java
/*
   Object creation of a class having private constructor
 */

package com.private_constructor_demo;

class Vehicle {

    private Vehicle(){
        System.out.println("constructor of Vehicle");
    }

    public static Vehicle getInstance() {
        return new Vehicle();
    }
}

public class Car {
    public static void main(String[] args) {
        Vehicle v = Vehicle.getInstance();
    }
}

/*
   constructor of Vehicle
 */

File: OOP-C++/Part3/Basics/2 Constructor and Destructor/Car.cpp
class Car : public Vehicle {

	public :
		int numGears;

	Car(int x, int y) : Vehicle(x) {		// Explicit call to Vehicle constructor.
		cout << "Car's parameterized constructor " << endl;
		numGears = y;
	}

	~Car() {
		cout << "Car's Destructor " << endl;
	}

	void print() {
		cout << "NumTyres : " << numTyres << endl;
		cout << "Color : " << color << endl;
		cout << "Num gears :" << numGears << endl;
	}
};


/*
	Implicit call to Vehicle default constructor look like.

	Car() : Vehicle() {
		cout << "Car's default constructor " << endl;
	}
	But if we use like this we are making explicit call to Vehicle's default constructor.
*/
